internal/npm_install/generate_build_file.ts

```diff
@@ -0,0 +1,1140 @@
+/**
+ * @license
+ * Copyright 2017 The Bazel Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * @fileoverview This script generates BUILD.bazel files by analyzing
+ * the node_modules folder layed out by yarn or npm. It generates
+ * fine grained Bazel `node_module_library` targets for each root npm package
+ * and all files for that package and its transitive deps are included
+ * in the target. For example, `@<workspace>//jasmine` would
+ * include all files in the jasmine npm package and all of its
+ * transitive dependencies.
+ *
+ * nodejs_binary targets are also generated for all `bin` scripts
+ * in each package. For example, the `@<workspace>//jasmine/bin:jasmine`
+ * target will be generated for the `jasmine` binary in the `jasmine`
+ * npm package.
+ *
+ * Additionally, a `@<workspace>//:node_modules` `node_module_library`
+ * is generated that includes all packages under node_modules
+ * as well as the .bin folder.
+ *
+ * This work is based off the fine grained deps concepts in
+ * https://github.com/pubref/rules_node developed by @pcj.
+ *
+ * @see https://docs.google.com/document/d/1AfjHMLVyE_vYwlHSK7k7yW_IIGppSxsQtPm9PTr1xEo
+ */
+'use strict';
+
+
+import * as fs from 'fs';
+import * as path from 'path';
+
+function log_verbose(...m: any[]) {
+  if (!!process.env['VERBOSE_LOGS']) console.error('[generate_build_file.js]', ...m);
+}
+
+const BUILD_FILE_HEADER = `# Generated file from yarn_install/npm_install rule.
+# See $(bazel info output_base)/external/build_bazel_rules_nodejs/internal/npm_install/generate_build_file.js
+
+# All rules in other repositories can use these targets
+package(default_visibility = ["//visibility:public"])
+
+`
+
+const args = process.argv.slice(2);
+const WORKSPACE = args[0];
+const RULE_TYPE = args[1];
+const ERROR_ON_BAZEL_FILES = parseInt(args[2]);
+const LOCK_FILE_LABEL = args[3];
+const INCLUDED_FILES = args[4] ? args[4].split(',') : [];
+const DYNAMIC_DEPS = JSON.parse(args[5] || '{}');
+
+if (require.main === module) {
+  main();
+}
+
+/**
+ * Create a new directory and any necessary subdirectories
+ * if they do not exist.
+ */
+function mkdirp(p: string) {
+  if (!fs.existsSync(p)) {
+    mkdirp(path.dirname(p));
+    fs.mkdirSync(p);
+  }
+}
+
+/**
+ * Writes a file, first ensuring that the directory to
+ * write to exists.
+ */
+function writeFileSync(p: string, content: any) {
```

Comment:
content should be `string`, right?

Reply:
the nodejs type expose it as any and i opted to propagate those though a case could have been made to only type how we use it.